flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ArrayIntersectFunction.java

```diff
@@ -0,0 +1,105 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime.functions.scalar;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.data.ArrayData;
+import org.apache.flink.table.data.GenericArrayData;
+import org.apache.flink.table.functions.BuiltInFunctionDefinitions;
+import org.apache.flink.table.functions.FunctionContext;
+import org.apache.flink.table.functions.SpecializedFunction;
+import org.apache.flink.table.runtime.util.EqualityAndHashcodeProvider;
+import org.apache.flink.table.runtime.util.ObjectContainer;
+import org.apache.flink.table.types.CollectionDataType;
+import org.apache.flink.table.types.DataType;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import javax.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+/** Implementation of {@link BuiltInFunctionDefinitions#ARRAY_INTERSECT}. */
+@Internal
+public class ArrayIntersectFunction extends BuiltInScalarFunction {
+    private final ArrayData.ElementGetter elementGetter;
+    private final EqualityAndHashcodeProvider equalityAndHashcodeProvider;
+
+    public ArrayIntersectFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.ARRAY_INTERSECT, context);
+        final DataType dataType =
+                ((CollectionDataType) context.getCallContext().getArgumentDataTypes().get(0))
+                        .getElementDataType()
+                        .toInternal();
+        elementGetter = ArrayData.createElementGetter(dataType.toInternal().getLogicalType());
+        this.equalityAndHashcodeProvider = new EqualityAndHashcodeProvider(context, dataType);
+    }
+
+    @Override
+    public void open(FunctionContext context) throws Exception {
+        equalityAndHashcodeProvider.open(context);
+    }
+
+    public @Nullable ArrayData eval(ArrayData arrayOne, ArrayData arrayTwo) {
+        try {
+            if (arrayOne == null || arrayTwo == null) {
+                return null;
+            }
+
+            Set<ObjectContainer> set = new HashSet<>();
+            for (int pos = 0; pos < arrayTwo.size(); pos++) {
+                final Object element = elementGetter.getElementOrNull(arrayTwo, pos);
+                final ObjectContainer objectContainer = createObjectContainer(element);
+                set.add(objectContainer);
+            }
+
+            Set<ObjectContainer> res = new LinkedHashSet<>();
```

Comment:
if a record present in set then add it to array list otherwise add it to set?
how to deduplicate?
if array1: a, a, array2: b, a
how to deduplicate a ,a ? @snuyanzin

Reply:
true